### PR TITLE
kubelet: fix nil pointer in startReflector for standalone mode

### LINKED ISSUE #113492: kubelet crashes when attempting to bring up a static pod that refers to a configmap
### What happened?

We were attempting to bring up a static pod on the node. This pod was attempting to mount a configmap using a volume:

```
    apiVersion: v1
    kind: Pod
    metadata:
      name: network-controller-manager
      namespace: kube-system
    spec:
      containers:
      - args:
        - --config=controller_manager_config.yaml
        command:
        - /manager
        .....
        volumeMounts:
        - mountPath: /controller_manager_config.yaml
          name: manager-config
          subPath: controller_manager_config.yaml
      hostNetwork: true
      volumes:
      - configMap:
          name: network-controller-manager
        name: manager-config
```
The configmap was as follows:
```
apiVersion: v1
    data:
      controller_manager_config.yaml: |
.....
    kind: ConfigMap
    name: network-controller-manager
    namespace: kube-system
```
We later realized that static pods don't support configMaps. However, when we attempted to run this cluster, kubelet kept crashing with this stack trace. We should probably fix this as the problem here is not obvious.

```
2022-10-31T20:00:48+0000 gke-b27bb59e05314f15b662-a7f3-3599-vm kubelet[3019]: E1031 20:00:48.166910 3019 runtime.go:79] Observed a panic: \"invalid memory address or nil pointer dereference\" (runtime error: invalid memory address or nil pointer dereference)\r\n
2022-10-31T20:00:48+0000 gke-b27bb59e05314f15b662-a7f3-3599-vm kubelet[3019]: goroutine 549 [running]:\r\n
2022-10-31T20:00:48+0000 gke-b27bb59e05314f15b662-a7f3-3599-vm kubelet[3019]: k8s.io/apimachinery/pkg/util/runtime.logPanic({0x3f89c00?, 0x71f2840})\r\n
2022-10-31T20:00:48+0000 gke-b27bb59e05314f15b662-a7f3-3599-vm kubelet[3019]: vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:75 +0x99\r\n
2022-10-31T20:00:48+0000 gke-b27bb59e05314f15b662-a7f3-3599-vm kubelet[3019]: k8s.io/apimachinery/pkg/util/runtime.HandleCrash({0x0, 0x0, 0x2540be400?})\r\n
2022-10-31T20:00:48+0000 gke-b27bb59e05314f15b662-a7f3-3599-vm kubelet[3019]: vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:49 +0x75\r\n
2022-10-31T20:00:48+0000 gke-b27bb59e05314f15b662-a7f3-3599-vm kubelet[3019]: panic({0x3f89c00, 0x71f2840})\r\n
2022-10-31T20:00:48+0000 gke-b27bb59e05314f15b662-a7f3-3599-vm kubelet[3019]: /usr/local/go/src/runtime/panic.go:884 +0x212\r\n
2022-10-31T20:00:48+0000 gke-b27bb59e05314f15b662-a7f3-3599-vm kubelet[3019]: k8s.io/client-go/tools/cache.(*Reflector).list(0xc000cbb770, 0xc000f85920)\r\n
2022-10-31T20:00:48+0000 gke-b27bb59e05314f15b662-a7f3-3599-vm kubelet[3019]: vendor/k8s.io/client-go/tools/cache/reflector.go:419 +0xb32\r\n
2022-10-31T20:00:48+0000 gke-b27bb59e05314f15b662-a7f3-3599-vm kubelet[3019]: k8s.io/client-go/tools/cache.(*Reflector).ListAndWatch(0xc000cbb770, 0xc000f85920)\r\n
2022-10-31T20:00:48+0000 gke-b27bb59e05314f15b662-a7f3-3599-vm kubelet[3019]: vendor/k8s.io/client-go/tools/cache/reflector.go:259 +0x152\r\n
2022-10-31T20:00:48+0000 gke-b27bb59e05314f15b662-a7f3-3599-vm kubelet[3019]: k8s.io/client-go/tools/cache.(*Reflector).Run.func1()\r\n
2022-10-31T20:00:48+0000 gke-b27bb59e05314f15b662-a7f3-3599-vm kubelet[3019]: vendor/k8s.io/client-go/tools/cache/reflector.go:223 +0x26\r\n
2022-10-31T20:00:48+0000 gke-b27bb59e05314f15b662-a7f3-3599-vm kubelet[3019]: k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1(0xc0011af000?)\r\n
2022-10-31T20:00:48+0000 gke-b27bb59e05314f15b662-a7f3-3599-vm kubelet[3019]: vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:157 +0x3e\r\n
2022-10-31T20:00:48+0000 gke-b27bb59e05314f15b662-a7f3-3599-vm kubelet[3019]: k8s.io/apimachinery/pkg/util/wait.BackoffUntil(0xc00092cc00?, {0x4fe2700, 0xc00075eaa0}, 0x1, 0xc000f85920)\r\n
2022-10-31T20:00:48+0000 gke-b27bb59e05314f15b662-a7f3-3599-vm kubelet[3019]: vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:158 +0xb6\r\n
2022-10-31T20:00:48+0000 gke-b27bb59e05314f15b662-a7f3-3599-vm kubelet[3019]: k8s.io/client-go/tools/cache.(*Reflector).Run(0xc000cbb770, 0xc000f85920)\r\n
2022-10-31T20:00:48+0000 gke-b27bb59e05314f15b662-a7f3-3599-vm kubelet[3019]: vendor/k8s.io/client-go/tools/cache/reflector.go:222 +0x17d\r\n
2022-10-31T20:00:48+0000 gke-b27bb59e05314f15b662-a7f3-3599-vm kubelet[3019]: k8s.io/kubernetes/pkg/kubelet/util/manager.(*objectCacheItem).startReflector(0xc000e03ce0)\r\n
2022-10-31T20:00:48+0000 gke-b27bb59e05314f15b662-a7f3-3599-vm kubelet[3019]: pkg/kubelet/util/manager/watch_based_manager.go:123 +0x91\r\n
2022-10-31T20:00:48+0000 gke-b27bb59e05314f15b662-a7f3-3599-vm kubelet[3019]: created by k8s.io/kubernetes/pkg/kubelet/util/manager.(*objectCache).newReflectorLocked\r\n
2022-10-31T20:00:48+0000 gke-b27bb59e05314f15b662-a7f3-3599-vm kubelet[3019]: pkg/kubelet/util/manager/watch_based_manager.go:243 +0x84b\r\n
2022-10-31T20:00:48+0000 gke-b27bb59e05314f15b662-a7f3-3599-vm kubelet[3019]: panic: runtime error: invalid memory address or nil pointer dereference [recovered]\r\n
2022-10-31T20:00:48+0000 gke-b27bb59e05314f15b662-a7f3-3599-vm kubelet[3019]: panic: runtime error: invalid memory address or nil pointer dereference\r\n
2022-10-31T20:00:48+0000 gke-b27bb59e05314f15b662-a7f3-3599-vm kubelet[3019]: goroutine 549 [running]:\r\n
2022-10-31T20:00:48+0000 gke-b27bb59e05314f15b662-a7f3-3599-vm kubelet[3019]: k8s.io/apimachinery/pkg/util/runtime.HandleCrash({0x0, 0x0, 0x2540be400?})\r\n
2022-10-31T20:00:48+0000 gke-b27bb59e05314f15b662-a7f3-3599-vm kubelet[3019]: vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:56 +0xd7\r\n
2022-10-31T20:00:48+0000 gke-b27bb59e05314f15b662-a7f3-3599-vm kubelet[3019]: panic({0x3f89c00, 0x71f2840})\r\n
2022-10-31T20:00:48+0000 gke-b27bb59e05314f15b662-a7f3-3599-vm kubelet[3019]: /usr/local/go/src/runtime/panic.go:884 +0x212\r\n
2022-10-31T20:00:48+0000 gke-b27bb59e05314f15b662-a7f3-3599-vm kubelet[3019]: k8s.io/client-go/tools/cache.(*Reflector).list(0xc000cbb770, 0xc000f85920)\r\n
2022-10-31T20:00:48+0000 gke-b27bb59e05314f15b662-a7f3-3599-vm kubelet[3019]: vendor/k8s.io/client-go/tools/cache/reflector.go:419 +0xb32\r\n
2022-10-31T20:00:48+0000 gke-b27bb59e05314f15b662-a7f3-3599-vm kubelet[3019]: k8s.io/client-go/tools/cache.(*Reflector).ListAndWatch(0xc000cbb770, 0xc000f85920)\r\n
2022-10-31T20:00:48+0000 gke-b27bb59e05314f15b662-a7f3-3599-vm kubelet[3019]: vendor/k8s.io/client-go/tools/cache/reflector.go:259 +0x152\r\n
2022-10-31T20:00:48+0000 gke-b27bb59e05314f15b662-a7f3-3599-vm kubelet[3019]: k8s.io/client-go/tools/cache.(*Reflector).Run.func1()\r\n
2022-10-31T20:00:48+0000 gke-b27bb59e05314f15b662-a7f3-3599-vm kubelet[3019]: vendor/k8s.io/client-go/tools/cache/reflector.go:223 +0x26\r\n
2022-10-31T20:00:48+0000 gke-b27bb59e05314f15b662-a7f3-3599-vm kubelet[3019]: k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1(0xc0011af000?)\r\n
2022-10-31T20:00:48+0000 gke-b27bb59e05314f15b662-a7f3-3599-vm kubelet[3019]: vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:157 +0x3e\r\n
2022-10-31T20:00:48+0000 gke-b27bb59e05314f15b662-a7f3-3599-vm kubelet[3019]: k8s.io/apimachinery/pkg/util/wait.BackoffUntil(0xc00092cc00?, {0x4fe2700, 0xc00075eaa0}, 0x1, 0xc000f85920)\r\n
2022-10-31T20:00:48+0000 gke-b27bb59e05314f15b662-a7f3-3599-vm kubelet[3019]: vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:158 +0xb6\r\n
2022-10-31T20:00:48+0000 gke-b27bb59e05314f15b662-a7f3-3599-vm kubelet[3019]: k8s.io/client-go/tools/cache.(*Reflector).Run(0xc000cbb770, 0xc000f85920)\r\n
2022-10-31T20:00:48+0000 gke-b27bb59e05314f15b662-a7f3-3599-vm kubelet[3019]: vendor/k8s.io/client-go/tools/cache/reflector.go:222 +0x17d\r\n
2022-10-31T20:00:48+0000 gke-b27bb59e05314f15b662-a7f3-3599-vm kubelet[3019]: k8s.io/kubernetes/pkg/kubelet/util/manager.(*objectCacheItem).startReflector(0xc000e03ce0)\r\n
2022-10-31T20:00:48+0000 gke-b27bb59e05314f15b662-a7f3-3599-vm kubelet[3019]: pkg/kubelet/util/manager/watch_based_manager.go:123 +0x91\r\n
2022-10-31T20:00:48+0000 gke-b27bb59e05314f15b662-a7f3-3599-vm kubelet[3019]: created by k8s.io/kubernetes/pkg/kubelet/util/manager.(*objectCache).newReflectorLocked\r\n
2022-10-31T20:00:48+0000 gke-b27bb59e05314f15b662-a7f3-3599-vm kubelet[3019]: pkg/kubelet/util/manager/watch_based_manager.go:243 +0x84b\r\n
2022-10-31T20:00:48+0000 gke-b27bb59e05314f15b662-a7f3-3599-vm systemd[1]: \x1b[0;1;39m\x1b[0;1;31m\x1b[0;1;39mkubelet.service: Main process exited, code=exited, status=2/INVALIDARGUMENT\x1b[0m\r\n
2022-10-31T20:00:48+0000 gke-b27bb59e05314f15b662-a7f3-3599-vm systemd[1]: \x1b[0;1;38;5;185m\x1b[0;1;39m\x1b[0;1;38;5;185mkubelet.service: Failed with result \'exit-code\'.\x1b[0m\r\n
```

### What did you expect to happen?

Instead of crashing Kubelet should have printed out a meaningful error message that would aid debugging.

### How can we reproduce it (as minimally and precisely as possible)?

Create a static pod that refers to a configMap and bring the pod up

### Anything else we need to know?

_No response_

### Kubernetes version

<details>

```console
$ kubectl version
Client Version: version.Info{Major:"1", Minor:"25", GitVersion:"v1.25.3", GitCommit:"434bfd82814af038ad94d62ebe59b133fcb50506", GitTreeState:"clean", BuildDate:"2022-10-12T10:57:26Z", GoVersion:"go1.19.2", Compiler:"gc", Platform:"linux/amd64"}
Kustomize Version: v4.5.7
Server Version: version.Info{Major:"1", Minor:"25", GitVersion:"v1.25.3-gke.900", GitCommit:"d7960fa8dcb507f46a99fb5f2ee0a1816cde8fea", GitTreeState:"clean", BuildDate:"2022-10-27T09:29:11Z", GoVersion:"go1.19.2 X:boringcrypto", Compiler:"gc", Platform:"linux/amd64"}
```

</details>

### Cloud provider

<details>
GKE
</details>

### OS version

<details>

```console
# On Linux:
$ cat /etc/os-release
NAME="Container-Optimized OS"
ID=cos
PRETTY_NAME="Container-Optimized OS from Google"
HOME_URL="https://cloud.google.com/container-optimized-os/docs"
BUG_REPORT_URL="https://cloud.google.com/container-optimized-os/docs/resources/support-policy#contact_us"
KERNEL_COMMIT_ID=6ffe18e56639e7815b07badca4d68fb0b54003cc
GOOGLE_METRICS_PRODUCT_ID=26
GOOGLE_CRASH_ID=Lakitu
VERSION=101
VERSION_ID=101
BUILD_ID=17162.40.13

$ uname -a
Linux gke-b5b62460724e4cfeae70-0376-801c-vm 5.15.65+ #1 SMP Sat Oct 15 09:12:54 UTC 2022 x86_64 Intel(R) Xeon(R) CPU @ 2.20GHz GenuineIntel GNU/Linux
```

</details>

### Install tools

<details>
None. Used GKE.
</details>

### Container runtime (CRI) and version (if applicable)

<details>

</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

### PR DESCRIPTION
**Pull request by pacoxu**
#### What type of PR is this?
/kind bug

#### What this PR does / why we need it:
1. The `startReflector` was introduced in https://github.com/kubernetes/kubernetes/pull/99393 @chenyw1990 since v1.21
2. https://github.com/kubernetes/kubernetes/pull/110138 fixed a goroutine leak in it @wojtek-t 
3. This PR focus on NPE fix.

#### Which issue(s) this PR fixes:

Fixes #113492

#### Special notes for your reviewer:
```
2022-10-31T20:00:48+0000 gke-b27bb59e05314f15b662-a7f3-3599-vm kubelet[3019]: k8s.io/kubernetes/pkg/kubelet/util/manager.(*objectCacheItem).startReflector(0xc000e03ce0)\r\n
2022-10-31T20:00:48+0000 gke-b27bb59e05314f15b662-a7f3-3599-vm kubelet[3019]: pkg/kubelet/util/manager/watch_based_manager.go:123 +0x91\r\n
```

https://github.com/kubernetes/kubernetes/blob/1f60f12e0ca7d26f9464e036b78f0f3974735ed4/pkg/kubelet/status/status_manager.go#L665-L672

if running in standalone mode, we should not update the pod status.

#### Does this PR introduce a user-facing change?
```release-note
kubelet: fix nil pointer in reflector start for standalone mode
```

### CODE REVIEW

**Comment by pacoxu** [comment]
BTW, should this be cherry-picked to v1.23-v1.25?

**Comment by aojea** [comment]
an e2e should be easy to add, is just creating a static pod with a reference to a configmap

https://github.com/kubernetes/kubernetes/issues/113492

File: pkg/kubelet/kubelet.go [hunk 1]
```diff
@@ -1635,7 +1635,7 @@ func (kl *Kubelet) syncPod(ctx context.Context, updateType kubetypes.SyncPodType
 	}
 
 	// ensure the kubelet knows about referenced secrets or configmaps used by the pod
-	if !kl.podWorkers.IsPodTerminationRequested(pod.UID) {
+	if !kl.podWorkers.IsPodTerminationRequested(pod.UID) && kl.kubeClient != nil {
 		if kl.secretManager != nil {
```
Comment (aojea) [inline pkg/kubelet/kubelet.go:1639]: question, why `kl.secretManager` and `kl.configMapManager` should be not nil in a standalone kubelet?

File: pkg/kubelet/status/status_manager.go [hunk 2]
```diff
@@ -666,6 +666,10 @@ func (m *manager) syncPod(uid types.UID, status versionedPodStatus) {
 		klog.V(1).InfoS("Status for pod is up-to-date; skipping", "podUID", uid)
 		return
 	}
+	// If kubeClient == nil, we are running in standalone mode
```
Comment (aojea) [inline pkg/kubelet/status/status_manager.go:669]: is this needed, I see that `Start()` returns early https://github.com/kubernetes/kubernetes/blob/9cf7b2e7a529aeaa9de9e20377b7a8a3dfa94d7f/pkg/kubelet/status/status_manager.go#L156-L163

is this called from another place?

File: pkg/kubelet/kubelet.go (hunk 1 above)
Comment (pacoxu) [reply pkg/kubelet/kubelet.go]: It seems to be better to not set secret and configmap manager if `kubeClient` is nil.

File: pkg/kubelet/status/status_manager.go (hunk 2 above)
Comment (pacoxu) [reply pkg/kubelet/status/status_manager.go]: I added this mistakenly. This should not work as you pointed out.

File: pkg/kubelet/volume_host.go [hunk 3]
```diff
@@ -257,11 +257,21 @@ func (kvh *kubeletVolumeHost) GetNodeAllocatable() (v1.ResourceList, error) {
 }
```
Comment (pacoxu) [inline pkg/kubelet/volume_host.go:258]: Volume Manager is needed. 
Or we can set the `not supproted` default func in `pkg/kubelet/kubelet.go`. Either is OK to me.

File: pkg/kubelet/kubelet.go [hunk 4]
```diff
@@ -550,24 +550,26 @@ func NewMainKubelet(kubeCfg *kubeletconfiginternal.KubeletConfiguration,
...
-	case kubeletconfiginternal.GetChangeDetectionStrategy:
-		secretManager = secret.NewSimpleSecretManager(kubeDeps.KubeClient)
-		configMapManager = configmap.NewSimpleConfigMapManager(kubeDeps.KubeClient)
-	default:
-		return nil, fmt.Errorf("unknown configmap and secret manager mode: %v", kubeCfg.ConfigMapAndSecretChangeDetectionStrategy)
-	}
-
-	klet.secretManager = secretManager
-	klet.configMapManager = configMapManager
+	if kubeDeps.KubeClient != nil {
```
Comment (bart0sh) [inline pkg/kubelet/kubelet.go:553]: Would it be more explicit to use klet.KubeClient instead of kubeDeps.KubeClient here and below?

Comment (pacoxu) [reply pkg/kubelet/kubelet.go]: Of course. They are the same in here.

File: pkg/kubelet/kubelet.go (hunk 1 above)
Comment (dims) [reply pkg/kubelet/kubelet.go]: yep, no one to talk to :)

**Comment by dims** [comment]
Anyway to test this @aojea @pacoxu ?

**Comment by pacoxu** [comment]
> Anyway to test this @aojea @pacoxu ?

I tested it locally with a hostnetwork pod with configmap in manifest path. No kube client related flag should be configured for kubelet is standalone mode.

**Comment by aojea** [comment]
> I tested it locally with a hostnetwork pod with configmap in manifest path. No kube client related flag should be configured for kubelet is standalone mode.


automated testing protects us for future regressions, we should automate everything if possible ... at least an unit test,

**Comment by pacoxu** [comment]
- with the ut add commit only: k8s.io/kubernetes/pkg/kubelet/kubelet.go (**49.7%)**: only test the NewMainKubelet behavior.

Before adding it, the coverage of the file is **30.5%**
https://storage.googleapis.com/kubernetes-jenkins/logs/ci-kubernetes-coverage-unit/1588342863123451904/artifacts/combined-coverage.html#file712

File: pkg/kubelet/kubelet.go [hunk 5]
```diff
@@ -181,7 +181,11 @@ const (
 	nodeLeaseRenewIntervalFraction = 0.25
 )
 
-var etcHostsPath = getContainerEtcHostsPath()
+var (
+	// ContainerLogsDir can be overwrited for testing usage
```
Comment (pacoxu) [inline pkg/kubelet/kubelet.go:185]: Only if we test `kubelet.Run()`, this is needed.

**Comment by aojea** [comment]
LGTM

@bobbypage @SergeyKanzhelev can you PTAL?

**Comment by aojea** [comment]
oh ,is still wip ?

**Comment by pacoxu** [comment]
Let me check again later.

**Comment by pacoxu** [comment]
> oh ,is still wip ?

I test on Mac directly at first and the previous CI failure is for  linus checking /dev/kmsg permission. Is it OK to add a fake check in linux oom code（a little odd）.

**Comment by pacoxu** [comment]
/test pull-kubernetes-unit 
to double check

**Comment by pacoxu** [comment]
ping @klueska 
Do you have time to take a look before code freeze?

File: pkg/kubelet/volume_host.go [hunk 6]
```diff
@@ -257,11 +257,21 @@ func (kvh *kubeletVolumeHost) GetNodeAllocatable() (v1.ResourceList, error) {
...
+	}
 }
 
 func (kvh *kubeletVolumeHost) GetConfigMapFunc() func(namespace, name string) (*v1.ConfigMap, error) {
-	return kvh.configMapManager.GetConfigMap
+	if kvh.configMapManager != nil {
+		return kvh.configMapManager.GetConfigMap
+	}
+	return func(namespace, name string) (*v1.ConfigMap, error) {
+		return nil, fmt.Errorf("not supported")
```
Comment (bobbypage) [inline pkg/kubelet/volume_host.go:273]: ditto here

File: pkg/kubelet/kubelet.go [hunk 7]
```diff
@@ -550,24 +554,26 @@ func NewMainKubelet(kubeCfg *kubeletconfiginternal.KubeletConfiguration,
...
-	case kubeletconfiginternal.GetChangeDetectionStrategy:
-		secretManager = secret.NewSimpleSecretManager(kubeDeps.KubeClient)
-		configMapManager = configmap.NewSimpleConfigMapManager(kubeDeps.KubeClient)
-	default:
-		return nil, fmt.Errorf("unknown configmap and secret manager mode: %v", kubeCfg.ConfigMapAndSecretChangeDetectionStrategy)
-	}
-
-	klet.secretManager = secretManager
-	klet.configMapManager = configMapManager
+	if klet.kubeClient != nil {
```
Comment (bobbypage) [inline pkg/kubelet/kubelet.go:557]: note: klet.kubeClient is equal to kubeDeps.KubeClient (set above on line 502)

**Comment by bobbypage** [comment]
LGTM, just a few doc comments, but change and test makes sense.

/lgtm

File: pkg/kubelet/kubelet.go (hunk 7 above)
Comment (pacoxu) [reply pkg/kubelet/kubelet.go:557]: yes.

https://github.com/kubernetes/kubernetes/pull/113501#discussion_r1011358876

- After `klet` is set, I changed to use `klet.kubeClient` here. 
- Before `klet` is set, we use `kubeDeps.KubeClient`

File: pkg/kubelet/volume_host.go [hunk 8]
```diff
@@ -257,11 +257,21 @@ func (kvh *kubeletVolumeHost) GetNodeAllocatable() (v1.ResourceList, error) {
 }
 
 func (kvh *kubeletVolumeHost) GetSecretFunc() func(namespace, name string) (*v1.Secret, error) {
-	return kvh.secretManager.GetSecret
+	if kvh.secretManager != nil {
+		return kvh.secretManager.GetSecret
+	}
+	return func(namespace, name string) (*v1.Secret, error) {
+		return nil, fmt.Errorf("not supported")
```
[the remark this answers is not in the document]
Comment (pacoxu) [reply pkg/kubelet/volume_host.go]: Updated.

File: pkg/kubelet/kubelet.go (hunk 7 above)
Comment (pacoxu) [reply pkg/kubelet/kubelet.go:557]: This is per @bart0sh 's comment.
- I think this is clear and will help if we will have some refactoring on this later.

**Review by mrunalp — APPROVED** [review]

**Comment by bobbypage** [comment]
Thanks for updates!

/lgtm

**Comment by pacoxu** [comment]
This is approved before code freeze. Can anyone help to add the milestone label?

@kubernetes/milestone-maintainers

**Pull request merged**